File: arrays/valid_palindrome/valid_palindrome.go
package main

import (
	"fmt"
	"unicode"
)

func main() {
	testCases := []struct {
		input    string
		expected bool
	}{
		{"A man, a plan, a canal: Panama", true},
		{"race a car", false},
		{" ", true},
		{"0P", false},
		{".,", true},
	}
	for _, tc := range testCases {
		result := isPalindrome(tc.input)
		status := "Passed"
		if result != tc.expected {
			status = "Failed"
		}
		fmt.Printf("%s isPalindrome(%q) = %v\n", status, tc.input, result)
	}
}

// function to check if given string is palindrome or not
func isPalindrome(s string) bool {
	l := 0
	r := len(s) - 1
	runes := []rune(s)
	for l < r {
		if !unicode.IsDigit(runes[l]) && !unicode.IsLetter(runes[l]) {
			l++
			continue
		}
		if !unicode.IsDigit(runes[r]) && !unicode.IsLetter(runes[r]) {
			r--
			continue
		}
		if unicode.ToLower(runes[l]) != unicode.ToLower(runes[r]) {
			return false
		}
		l++
		r--
	}
	return true
}

File: week2/valid_palindrome/valid_palindrome.py
def is_palindrome(s: str) -> bool:
    l = 0
    r = len(s) - 1
    while l < r:
        # skip non-alphanumeric on left
        if not s[l].isalnum():
            l += 1
            continue
        # skip non-alphanumeric on right
        if not s[r].isalnum():
            r -= 1
            continue
        # compare characters case-insensitively
        if s[l].lower() != s[r].lower():
            return False
        l += 1
        r -= 1

    return True

if __name__ == "__main__":
    tests = [
        ("A man, a plan, a canal: Panama", True),
        ("race a car", False),
        (" ", True),
        ("0P", False),
        (".,", True),
    ]
    for s, expected in tests:
        result = is_palindrome(s)
        status = "Passed" if result == expected else "Failed"
        print(f'{status} is_palindrome("{s}") = {result}')